src/network/networkd-neighbor.c

```diff
@@ -0,0 +1,215 @@
+                                          link->ifindex, neighbor->family);
+        if (r < 0)
+                return log_error_errno(r, "Could not allocate RTM_NEWNEIGH message: %m");
+
+        r = sd_rtnl_message_neigh_set_state(req, NUD_PERMANENT);
+        if (r < 0)
+                return log_error_errno(r, "Could not set state: %m");
+
+        r = sd_netlink_message_set_flags(req, NLM_F_REQUEST | NLM_F_CREATE | NLM_F_REPLACE);
+        if (r < 0)
+                return log_error_errno(r, "Could not set flags: %m");
+
+        r = sd_netlink_message_append_ether_addr(req, NDA_LLADDR, &neighbor->mac);
+        if (r < 0)
+                return log_error_errno(r, "Could not append NDA_LLADDR attribute: %m");
+
+        switch (neighbor->family) {
+        case AF_INET6:
+                r = sd_netlink_message_append_in6_addr(req, NDA_DST, &neighbor->in_addr.in6);
+                if (r < 0)
+                        return log_error_errno(r, "Could not append NDA_DST attribute: %m");
+                break;
+        case AF_INET:
+                r = sd_netlink_message_append_in_addr(req, NDA_DST, &neighbor->in_addr.in);
+                if (r < 0)
+                        return log_error_errno(r, "Could not append NDA_DST attribute: %m");
+                break;
+        default:
+                return log_error_errno(-EINVAL, "Invalid neighbor address type");
+        }
+
+        r = sd_netlink_call_async(link->manager->rtnl, NULL, req, callback,
+                                  link_netlink_destroy_callback, link, 0, __func__);
+        if (r < 0)
+                return log_error_errno(r, "Could not send rtnetlink message: %m");
+
+        link_ref(link);
+
+        return 0;
+}
+
+int config_parse_neighbor_address(const char *unit,
+                                  const char *filename,
+                                  unsigned line,
+                                  const char *section,
+                                  unsigned section_line,
+                                  const char *lvalue,
+                                  int ltype,
+                                  const char *rvalue,
+                                  void *data,
+                                  void *userdata) {
+
+        Network *network = userdata;
+        _cleanup_(neighbor_freep) Neighbor *n = NULL;
+        union in_addr_union buffer;
+        int r, f;
+
+        assert(filename);
+        assert(section);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        r = neighbor_new_static(network, filename, section_line, &n);
+        if (r < 0)
+                return r;
+
+        r = in_addr_from_string_auto(rvalue, &f, &buffer);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, r, "Neighbor Address is invalid, ignoring assignment: %s", rvalue);
+                return 0;
+        }
+
+        n->family = f;
+        n->in_addr = buffer;
```

Comment:
why the indirection through f/buffer? in_addr_from_string_auto() could directly write to these fields of `n`, no?

Reply:
The main reason is that the underlying function inet_pton(3) does not guarantee that it won't overwrite the buffer on failure. If we specify two addresses in the config file and the second one is bad and inet_pton overwrites the buffer, we won't end up with a bad address in the output `struct Neighbor`. It does seem silly that inet_pton would clobber the `dst` when an error occurs, but I'm just conforming to the pattern used in other places. An example of similar behavior is in src/networkd/networkd-route.

Reply:
> The main reason is that the underlying function inet_pton(3) does not guarantee that it won't overwrite the buffer on failure.

https://github.com/kraj/glibc/blob/02cd5c1a8d033d7f91fea12a66bb44d1bbf85f76/resolv/inet_pton.c#L75-L78
https://github.com/kraj/glibc/blob/02cd5c1a8d033d7f91fea12a66bb44d1bbf85f76/resolv/inet_pton.c#L138-L140

Reply:
@wkennington `in_addr_from_string_auto()` really should be abstracting the weirdness of inet_pton() away. It is in a way a typesafe version of it, after all.

Reply:
Done